src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/CommonInterests.cs

```diff
@@ -0,0 +1,201 @@
+            for (int i = namespaces.Count - 1; i >= 0; i--)
+            {
+                if (currentNamespace?.Name != namespaces[i])
+                {
+                    return false;
+                }
+
+                currentNamespace = currentNamespace.ContainingNamespace;
+            }
+
+            return currentNamespace?.IsGlobalNamespace ?? false;
+        }
+
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+        internal readonly struct QualifiedType
+#pragma warning restore CA1815 // Override equals and operator equals on value types
+        {
+            public QualifiedType(IReadOnlyList<string> containingTypeNamespace, string typeName)
+            {
+                this.Namespace = containingTypeNamespace;
+                this.Name = typeName;
+            }
+
+            public IReadOnlyList<string> Namespace { get; }
+
+            public string Name { get; }
+
+            public bool IsMatch(ISymbol symbol)
+            {
+                return symbol?.Name == this.Name
+                    && symbol.BelongsToNamespace(this.Namespace);
+            }
+
+            public override string ToString() => string.Join(".", this.Namespace.Concat(new[] { this.Name }));
+        }
+
+        internal static class AsyncMethodBuilderAttribute
+        {
+            internal const string TypeName = nameof(System.Runtime.CompilerServices.AsyncMethodBuilderAttribute);
+            internal static readonly IReadOnlyList<string> Namespace = Namespaces.SystemRuntimeCompilerServices;
+        }
+
+        internal static bool HasAsyncCompatibleReturnType([NotNullWhen(true)] this IMethodSymbol? methodSymbol) => IsAsyncCompatibleReturnType(methodSymbol?.ReturnType);
+
+        /// <summary>
+        /// Determines whether a type could be used with the async modifier as a method return type.
+        /// </summary>
+        /// <param name="typeSymbol">The type returned from a method.</param>
+        /// <returns><c>true</c> if the type can be returned from an async method.</returns>
+        /// <remarks>
+        /// This is not the same thing as being an *awaitable* type, which is a much lower bar. Any type can be made awaitable by offering a GetAwaiter method
+        /// that follows the proper pattern. But being an async-compatible type in this sense is a type that can be returned from a method carrying the async keyword modifier,
+        /// in that the type is either the special Task type, or offers an async method builder of its own.
+        /// </remarks>
+        internal static bool IsAsyncCompatibleReturnType([NotNullWhen(true)] this ITypeSymbol? typeSymbol)
+        {
+            if (typeSymbol is null)
+            {
+                return false;
+            }
+
+            // ValueTask and ValueTask<T> have the AsyncMethodBuilderAttribute
+            return (typeSymbol.Name == nameof(Task) && typeSymbol.BelongsToNamespace(Namespaces.SystemThreadingTasks))
```

Comment:
The project already has definitions for all the task types in WellKnownTypes.  The analyzer should fetch those once at the beginning and then just compare the symbols.  You can look at other task-related analyzers in the repo to see how they do it, e.g.
https://github.com/dotnet/roslyn-analyzers/blob/086d1cc888ca0d8c96b94444935f488fcaf3fe98/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/DoNotDirectlyAwaitATask.cs#L144